### ISSUE TITLE
Add regularization loss in build_loss_and_gradients?

### ISSUE THREAD

**Issue by taochenshh** [blei-lab/edward#529]
Current implementation of loss in [variational_inference.py](https://github.com/blei-lab/edward/blob/master/edward/inferences/variational_inference.py#L64) does not include  `ops.GraphKeys.LOSSES` and `ops.GraphKeys.REGULARIZATION_LOSSES` in tensorflow graph. So I cannot add regularization loss to the loss used in update. Of course, I can manually add it like `inference.loss += tf.losses.get_total_loss()` after calling `inference.initialize(var_list=tf.trainable_variables())`, but that means I have to modify `inference.train` op as well. Will edward support adding regularization loss in the future release?

**Comment by dustinvtran** [comment]
Great question. I guess regularization losses reaches a middleground, between leveraging losses/gradients from existing algorithms and just inheriting classes and writing your own `build_loss_and_gradients` method. It sounds like a useful feature to add on top to all existing variational algorithms. Contributions welcome.

From my understanding, the implementation is simple. We simply scope out the ops from those GraphKeys; then add it to the various losses we work with before calling `tf.gradients`.

**Comment by fujiisoup** [comment]
I am also interested in it.
Addition to the regularization, I sometimes want to clip gradient to stabilize the learning.

It would be great if we could access loss or its gradients in some form.
API I am thinking of is like
```python
def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
                 gradient_generator=None,
                 global_step=None, *args, **kwargs):
```
 with 
```python
def gradient_generator(loss, grad, v):
  loss = loss + reg * tf.square(v)
  return loss, tf.gradient(loss, v)
```
or
```python
def gradient_generator(loss, grad, v):
  return loss, tf.clip_by_value(tf.gradient(loss, v, -0.1, 0.1))
```

Does this API look intuitive?
If this modification looks general enough, I will try to contribute.

**Comment by dustinvtran** [comment]
Thanks for the proposal. Can you add it as a separate issue? It sounds unrelated to this one. I'll add thoughts on that new thread.

**Comment by fujiisoup** [comment]
@dustinvtran 
Thanks. I raise another issue in #708.

**Comment by davharris** [comment]
This is a hack, but it seems like it should provide a workaround until the API is changed.  It lets you add an arbitrary loss/regularizer via a custom random variable.

```
# Increment the loss by some value (`loss`)
# Most of this boilerplate is probably unnecessary, but the docs are out-of-date and so I 
# copy-pasted from https://discourse.edwardlib.org/t/simple-wrapper-custom-variable/367/4
class distributions_Penalty(Distribution):
    def __init__(self, loss,
                  validate_args=False,
                  allow_nan_stats=True,
                  name="Penalty"):
        self._loss = loss # Store the loss we want to use
        super(distributions_Penalty, self).__init__(
            dtype=tf.float32,
            reparameterization_type=tf.contrib.distributions.FULLY_REPARAMETERIZED,
            validate_args=validate_args,
            allow_nan_stats=allow_nan_stats,
            name=name)

    def _log_prob(self, value):
        # This is the key line. The "log_prob" is just the negative penalty.
        return -self._loss

    def _sample_n(self, n, seed=None):
        # This should never be called by the user; it's necessary so Edward doesn't complain
        return tf.zeros(n)

class Penalty(RandomVariable, distributions_Penalty):
    def __init__(self, *args, **kwargs):
        RandomVariable.__init__(self, *args, **kwargs)
```

Once `Penalty` is defined, you can just:

1) add `penalty = Penalty(extra_loss)` to the model

2) list `penalty: np.nan` (or any other value) in the `data` dictionary for `inference`.

**Comment by dustinvtran** [comment]
ad hoc loss terms and gradient clipping are excellent examples that I think test the limitations of Edward's inference design. I'm not totally confident with a solution that not just works but is very natural. E.g., defining clipped optimizer classes and penalty-distribution classes are great temporary solutions. But you would hope taht Edward has a more puritanical solution where penalties and clipping are part of the training process.

**Comment by siddharth-agrawal** [comment]
@dustinvtran I miscommunicated with my PR. There are still other files to be modified, I wanted to cover just `klqp.py` in one PR because it would become too long with all the changes. This issue should still be open.

**Comment by dustinvtran** [comment]
I'm not sure how this one got auto-closed but the other ones did not.. Re-opening.

**Comment by siddharth-agrawal** [comment]
@dustinvtran I figured out why it was automatically closed. In PR #813 I mentioned the word "fixes" which has been underlined. When you hover over it, the box says "this PR closes issue #529".

**Comment by siddharth-agrawal** [comment]
@dustinvtran I guess this should be closed. :)

**Comment by dustinvtran** [comment]
+1
